src/Rewards.sol

```diff
@@ -0,0 +1,239 @@
+     */
+    NODL public nodlToken;
+
+    /**
+     * @dev Maximum amount of rewards that can be distributed in a period.
+     */
+    uint256 public rewardQuota;
+
+    /**
+     * @dev Duration of each reward period.
+     */
+    uint256 public rewardPeriod;
+
+    /**
+     * @dev Timestamp indicating when the reward quota is due to be renewed.
+     */
+    uint256 public quotaRenewalTimestamp;
+
+    /**
+     * @dev Amount of rewards claimed in the current period.
+     */
+    uint256 public rewardsClaimed;
+
+    /**
+     * @dev Address of the authorized oracle.
+     */
+    address public authorizedOracle;
+
+    /**
+     * @dev Mapping to store reward sequences for each recipient to prevent replay attacks.
+     */
+    mapping(address => uint256) public rewardSequences;
+
+    /**
+     * @dev Struct on which basis an individual reward must be issued.
+     */
+    struct Reward {
+        address recipient;
+        uint256 amount;
+        uint256 sequence;
+    }
+
+    /**
+     * @dev Error when the reward quota is exceeded.
+     */
+    error RewardQuotaExceeded();
+
+    /**
+     * @dev Error indicating the reward renewal period is set to zero which is not acceptable.
+     */
+    error ZeroPeriod();
+
+    /**
+     * @dev Error indicating that scheduling the reward quota renewal has failed most likley due to the period being too long.
+     */
+    error TooLongPeriod();
+
+    /**
+     * @dev Error when the reward is not from the authorized oracle.
+     */
+    error UnauthorizedOracle();
+
+    /**
+     * @dev Error when the recipient's reward sequence does not match.
+     */
+    error InvalidRecipientSequence();
+
+    /**
+     * @dev Event emitted when the reward quota is set.
+     */
+    event RewardQuotaSet(uint256 quota);
+
+    /**
+     * @dev Event emitted when a reward is minted.
+     */
+    event RewardMinted(address indexed recipient, uint256 amount, uint256 totalRewardsClaimed);
+
+    /**
+     * @dev Initializes the contract with the specified parameters.
+     * @param token Address of the NODL token contract.
+     * @param initialQuota Initial reward quota.
+     * @param initialPeriod Initial reward period.
+     * @param oracleAddress Address of the authorized oracle.
+     */
+    constructor(NODL token, uint256 initialQuota, uint256 initialPeriod, address oracleAddress)
+        EIP712(SIGNING_DOMAIN, SIGNATURE_VERSION)
+    {
+        // This is to avoid the ongoinb overhead of safe math operations
+        if (initialPeriod == 0) {
+            revert ZeroPeriod();
+        }
+        // This is to prevent overflows while avoiding the ongoing overhead of safe math operations
+        if (initialPeriod > MAX_PERIOD) {
+            revert TooLongPeriod();
+        }
+
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        nodlToken = token;
+        rewardQuota = initialQuota;
+        rewardPeriod = initialPeriod;
+        rewardsClaimed = 0;
```

Comment:
gas optimization: `rewardsClaimed` is already 0, there is no need to double initialize it

Reply:
These should be optimized by the compiler, I would be surprised if not. But no problem to change it.